beyond-all-reason/spring
# Add directional pathing

Fixes https://github.com/beyond-all-reason/spring/issues/1398

Includes fixes from https://github.com/beyond-all-reason/spring/pull/1395 which Fixes: https://github.com/beyond-all-reason/spring/issues/1386

This change provides the capability to use directional pathing (HAFPS.) I.e. the result of a pathing query takes the direction of the search into account. So you can path down a hill you wouldn't be able to travel up for example.

This change also adds the capability to use shortest route path (QTPFS.) Normally paths are given according to the fastest route. Now it is possible to have units take the shorest route, which helps them follow the expected path of players better - for example, avoiding dangerous areas.

To use directional pathing:

1. Ensure modrule: `movement.allowDirectionalPathing = true`, it is `true` by default.
2. Use a compatible pathing system. Only HAPFS is compatible at the moment. `system.pathFinderSystem = 0`. it is `0` by default.
3. Enable/disable `allowDirectionalPathing` field on any MoveDefs as you see fit. It is enabled by default.

Note: using an incompatible pathing system (like QTPFS) will disable directional pathing, regardless of other settings.
Note: turning off `allowDirectionalPathing` in modrules will also disable directional pathing, regardless of other settings.


To use shortest route pathing:

1. Ensure modrule: `movement.preferShortestPath= true`, it is `false` by default.
2. Use a compatible pathing system. Only QTPFS is compatible at the moment. `system.pathFinderSystem = 1`. it is `0` by default.
3. Enable/disable `preferShortestPath` field on any MoveDefs as you see fit. It is enabled by default.

Note: using an incompatible pathing system (like HAPFS) will disable shortest route pathing, regardless of other settings.
Note: turning off `preferShortestPath` in modrules will also disable shortest route pathing, regardless of other settings.


Notes:

Using directional pathing has a performance COST on HAPFS.
Using shortest route pathing has a performance BENEFIT on QTPFS.
